integration-tests/chaos/congestion/simulator.go

```diff
@@ -0,0 +1,484 @@
+	t.Cleanup(func() {
+		_ = s.Close()
+	})
+
+	return s, nil
+}
+
+type phaseType int
+
+const (
+	phaseInactive phaseType = iota
+	phaseRampUp
+	phasePlateau
+	phaseCoolDown
+)
+
+type chainState struct {
+	PhaseName           phaseType
+	Congestion          float64
+	TipDeltaPercent     float64
+	BaseFeeDeltaPercent float64
+}
+
+func (s *Simulator) start(ctx context.Context) error {
+	if !s.Input.Enabled {
+		s.lggr.Infow("Congestion simulator is not enabled - won't run")
+		return nil
+	}
+	s.expectedCurrentCongestion.Store(float64(0))
+	s.expectedCurrentTip.Store(s.InitialTipPerGas)
+
+	if s.InitialBaseFeePerGas == 0 {
+		header, err := s.client.Client.HeaderByNumber(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("failed to get header by number to capture initial base fee: %w", err)
+		}
+
+		s.InitialBaseFeePerGas = header.BaseFee.Uint64()
+		s.lggr.Debugf("InitialBaseFeePerGas was not provided using: %d", s.InitialBaseFeePerGas)
+	}
+
+	s.expectedCurrentBaseFee.Store(s.InitialBaseFeePerGas)
+
+	// deploy logs emitter that we'll use to fill blocks
+	var err error
+	s.emitter, err = contracts.DeployLogEmitterContract(logging.GetTestLogger(s.t), s.client)
+	if err != nil {
+		return fmt.Errorf("failed to create emitter contract: %w", err)
+	}
+
+	if s.Input.PayloadSize == nil {
+		payloadSize, err := s.estimatePayloadSizeToFillBlock(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to estimate payload size to fill block: %w", err)
+		}
+		s.Input.PayloadSize = &payloadSize
+	}
+	s.payload = []string{string(make([]byte, *s.Input.PayloadSize))}
+
+	if s.Input.GasLimit == 0 {
+		gasLimit, err := s.estimateTxGasLimit(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to estimate gas limit: %w", err)
+		}
+		s.Input.GasLimit = gasLimit
+	}
+
+	ticker := services.TickerConfig{Initial: time.Second * time.Duration(s.InitialDelay)}.NewTicker(time.Second * time.Duration(s.Period))
+	s.eng.GoTick(ticker, func(ctx context.Context) {
+		s.runTick(ctx)
+		ticker.Reset() // ensure that each iteration is delayed by s.Period
+	})
+	// report observations
+	s.eng.Go(func(ctx context.Context) {
+		s.listenHeadsUntilCanceled(ctx, s.reportObservationOnNewHead)
+	})
+	return nil
+}
+
+func (s *Simulator) startTransactionProducer(ctx context.Context, wg *sync.WaitGroup) {
+	numberOfSenders := s.NumberOfTxsPerBlock
+	work := make(chan struct{}, numberOfSenders)
+	wg.Add(numberOfSenders)
+	for i := range numberOfSenders {
+		go s.runSender(ctx, wg, i, work)
+	}
+
+	txsToSend := numberOfSenders / 2
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.listenHeadsUntilCanceled(ctx, func(ctx context.Context, head *types.Header) error {
+			congestion := float64(head.GasUsed) / float64(head.GasLimit)
+			congestionTarget := s.expectedCurrentCongestion.Load().(float64)
+			if txsToSend < numberOfSenders && congestion < congestionTarget*0.95 {
+				txsToSend++
+			} else if txsToSend > 0 && congestion > congestionTarget*1.05 {
+				txsToSend--
+			}
```

Comment:
comment why `*0.95` and `*1.05`?

could these values need to be adjusted if block time changes?

Reply:
Nice catch. No, it should not depend on the block time. Initially, I've introduced it to factor in rounding errors and to make the adjustments less aggressive. Now seems like other improvements made the aggressiveness issue less severe, so we can drop the error factor from 5% to 1%.